compiler/rustc_mir_transform/src/simplify_pow_of_two.rs

```diff
@@ -0,0 +1,212 @@
+//! A pass that checks for and simplifies calls to `pow` where the receiver is a power of
+//! two. This can be done with `<<` instead.
+
+use crate::MirPass;
+use rustc_const_eval::interpret::{ConstValue, Scalar};
+use rustc_middle::mir::patch::MirPatch;
+use rustc_middle::mir::*;
+use rustc_middle::ty::{self, Ty, TyCtxt, UintTy};
+use rustc_span::sym;
+use rustc_target::abi::FieldIdx;
+
+pub struct SimplifyPowOfTwo;
+
+impl<'tcx> MirPass<'tcx> for SimplifyPowOfTwo {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        let mut patch = MirPatch::new(body);
+
+        for (i, bb) in body.basic_blocks.iter_enumerated() {
+            let term = bb.terminator();
+            let source_info = term.source_info;
+            let span = source_info.span;
+
+            if let TerminatorKind::Call {
+                    func,
+                    args,
+                    destination,
+                    target: Some(target),
+                    call_source: CallSource::Normal,
+                    ..
+                } = &term.kind
+                && let Some(def_id) = func.const_fn_def().map(|def| def.0)
+                && let def_path = tcx.def_path(def_id)
+                && tcx.crate_name(def_path.krate) == sym::core
+                && let [recv, exp] = args.as_slice()
+                && let Some(recv_const) = recv.constant()
+                && let ConstantKind::Val(
+                    ConstValue::Scalar(Scalar::Int(recv_int)),
+                    recv_ty,
+                ) = recv_const.literal
+                && recv_ty.is_integral()
+                && tcx.item_name(def_id) == sym::pow
```

Comment:
This detection mecanism is brittle. The compiler should be as independent as possible from the exact paths in core. Can diagnostic items be used for this ?

Reply:
I'm ok with it, but if so we should use lang items instead as pointed out by https://github.com/rust-lang/rust/pull/114254#discussion_r1278605177. But this feels like it'd be very verbose, as we'd need 12 lang items for them all and then check whether it's any (rather than just if the item name is `pow`).